Probability Samplers based on W3C Trace Context Level 2 (#3910)
open-telemetry/opentelemetry-specification: specification/trace/sdk.md

#### TraceIdRatioBased

The `TraceIdRatioBased` MUST ignore the parent `SampledFlag`. To respect the
parent `SampledFlag`, the `TraceIdRatioBased` should be used as a delegate of
the `ParentBased` sampler specified below.

##### `TraceIdRatioBased` sampler implementation overview

SDKs SHOULD use an implementation that is equivalent to the
specification referenced below as the `TraceIdRatioBased` Sampler,
which is based on the [W3C Trace Context Level 2 Random TraceID
flag](https://www.w3.org/TR/trace-context-2/#random-trace-id-flag).
When a TraceID has the `Random` bit set, samplers are able to use 56
specific bits of consistent randomness for sampling decisions.

The implementation has the following steps:

* Ratio values are restricted to the range `2**-56` through 1
* A rejection threshold is calculated, expressing as an integer how

>   many out of 2**56 trace IDs should be sampled.

Reviewer: "sampled" => "dropped"

> * A rejection threshold is calculated, expressing as an integer how
>   many out of 2**56 trace IDs should be sampled.

Reviewer 2: ```suggestion
* A rejection threshold is calculated, expressing as an integer the number
  out of 2**56 trace IDs that should be dropped.
```

* The threshold is encoded as a "T-value", expressing the threshold
  using up to 5 hexadecimal digits of precision.
* Sampler decisions are made by comparing the Trace ID randomness
  against the rejection threshold.

> * When Sampled, T-value is included in the [OpenTelemetry TraceState

Reviewer 3: It will be good rephrase this to an active voice of what implementations must do. For example, something like:  "When sampled, an implementation MUST update the tracestate header to specify the T-value. <+ any additional details>"

>   header](./tracestate-handling.md), identified by sub-key `th`,

Reviewer 3: Should we update the tracestate-handling.md to include the examples for the sub-key 'th'?  Right now, it seems to be referring to the earlier p values and r value examples.

X-Ray trace id generator MUST NOT be maintained or distributed as part of the
Core OpenTelemetry repositories.

### Randomness requirements

The SDK SHOULD implement the TraceID randomness requirements specified
in the W3C [Trace Context Level
2](https://www.w3.org/TR/trace-context-2/#randomness-of-trace-id)
Candidate Recommendation.

> This states that the SDK should fill least significant 7 bytes (i.e., 56
> bits) of the TraceID are genuinely random or pseudo-random., so they
> can be used for probability sampling.

Reviewer 2: ```suggestion
This states that the SDK should fill the least significant 7 bytes (i.e., 56
bits) of the TraceID with bits that are genuinely random or pseudo-random, 
so that they can be used for probability sampling.
```
